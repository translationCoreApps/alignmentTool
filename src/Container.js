import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {DragDropContext} from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import WordList from './components/WordList';
import AlignmentGrid from './components/AlignmentGrid';
import isEqual from 'lodash/isEqual';
import {getWords, getAlignedWords, disableAlignedWords} from './utils/words';

/**
 * The base container for this tool
 */
class Container extends Component {

  componentWillMount() {
    // current panes persisted in the scripture pane settings.
    const {ScripturePane} = this.props.settingsReducer.toolsSettings;
    let panes = [];
    if (ScripturePane) panes = ScripturePane.currentPaneSettings;
    // filter out targetLanguage and bhp
    panes = panes.filter((pane) => {
      return pane !== 'targetLanguage' && pane !== 'bhp' && pane !== 'ugnt';
    });
    // getting the last pane from the panes array if it exist otherwise equal to null.
    const lastPane = panes[panes.length - 1] ? panes[panes.length - 1] : null;
    // set the ScripturePane to display targetLanguage and bhp for the word alignment tool from left to right.
    let desiredPanes = ['targetLanguage', 'ugnt'];
    // if last pane found in previous scripture pane settings then carry it over to new settings in wordAlignment.
    if (lastPane && lastPane !== 'targetLanguage' && lastPane !== 'bhp' &&
      lastPane !== 'ugnt') desiredPanes.push(lastPane);
    // set new pane settings
    this.props.actions.setToolSettings('ScripturePane', 'currentPaneSettings',
      desiredPanes);
  }

  componentWillReceiveProps(nextProps) {
    if (!isEqual(this.props.contextIdReducer.contextId,
        nextProps.contextIdReducer.contextId)) {
      let page = document.getElementById('AlignmentGrid');
      if (page) page.scrollTop = 0;
    }
  }

  render() {
    // Modules not defined within translationWords
    const {ScripturePane} = this.props.currentToolViews;
    let scripturePane = <div/>;
    // populate scripturePane so that when required data is preset that it renders as intended.
    if (Object.keys(this.props.resourcesReducer.bibles).length > 0) {
      scripturePane = <ScripturePane {...this.props} />;
    }

    const {moveBackToWordBank} = this.props.actions;
    const {connectDropTarget, isOver} = this.props;
    const {alignmentData} = this.props.wordAlignmentReducer;
    const {contextId} = this.props.contextIdReducer;
    const {lexicons, bibles: {ugnt, targetLanguage}} = this.props.resourcesReducer;
    let chapter, verse;
    let words = [];
    if (contextId) {
      chapter = contextId.reference.chapter;
      verse = contextId.reference.verse;

      // parse secondary text words
      const secondaryChapterText = targetLanguage[chapter];
      words = getWords(secondaryChapterText[verse]);
      const alignedWords = getAlignedWords(alignmentData, chapter, verse);
      words = disableAlignedWords(words, alignedWords);
    }

    return (
      <div style={{display: 'flex', width: '100%', height: '100%'}}>
        <WordList chapter={chapter}
                  verse={verse}
                  words={words}
                  moveBackToWordBank={moveBackToWordBank}
                  connectDropTarget={connectDropTarget}
                  isOver={isOver}/>
        <div style={{
          flex: 0.8,
          width: '100%',
          height: '100%',
          paddingBottom: '150px'
        }}>
          {scripturePane}
          <AlignmentGrid alignmentData={alignmentData}
                         lexicons={lexicons}
                         actions={this.props.actions}
                         contextId={contextId}/>
        </div>
      </div>
    );
  }
}

Container.propTypes = {
  currentToolViews: PropTypes.object.isRequired,
  resourcesReducer: PropTypes.object.isRequired,
  contextIdReducer: PropTypes.object.isRequired,
  settingsReducer: PropTypes.object.isRequired,
  wordAlignmentReducer: PropTypes.object.isRequired,
  actions: PropTypes.object.isRequired,

  isOver: PropTypes.bool,
  connectDropTarget: PropTypes.func
};

export default DragDropContext(HTML5Backend)(Container);
