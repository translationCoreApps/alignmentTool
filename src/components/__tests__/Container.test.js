/* eslint-env jest */
import React, {Component} from 'react';
import TestBackend from 'react-dnd-test-backend';
import {DragDropContext} from 'react-dnd';
import expect from 'expect';
import Container from '../Container';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import {connectTool} from 'tc-tool';

test('Container renders', () => {
  const props = {
    actions: {
      setToolSettings: jest.fn(),
      getWordListForVerse: jest.fn(),
      loadLexiconEntry: jest.fn(),
      showPopover: jest.fn()
    },
    settingsReducer: {
      toolsSettings: {
        ScripturePane: {
          currentPaneSettings: []
        }
      }
    },
    selectionsReducer: {
      selections: [{text: 'text'}]
    },
    projectDetailsReducer: {},
    contextIdReducer: {},
    currentToolViews: {},
    resourcesReducer: {
      bibles: [],
      lexicons: {}
    },
    wordAlignmentReducer: {
      alignmentData: {}
    },
    appLanguage: 'en',
    tc: {
      sourceChapter: {},
      targetChapter: {},
      showDialog: jest.fn(),
      writeGlobalToolData: jest.fn(),
      readGlobalToolData: jest.fn(),
      appLanguage: 'en',
      showLoading: jest.fn(),
      closeLoading: jest.fn()
    },
    translate: k => k
  };

  const WrappedContainer = wrapInTestContext(Container);
  const wrapper = mount(<WrappedContainer {...props} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

/**
 * Wraps a component into a DragDropContext that uses the TestBackend.
 */
function wrapInTestContext(DecoratedComponent) {
  class TestContextContainer extends Component {
    render() {
      return <DecoratedComponent {...this.props} />;
    }
  }

  return DragDropContext(TestBackend)(connectTool()(TestContextContainer).container);
}
