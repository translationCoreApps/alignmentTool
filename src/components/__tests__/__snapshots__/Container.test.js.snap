// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container renders 1`] = `
<DragDropContext(Tool)
  actions={
    Object {
      "getWordListForVerse": [Function],
      "loadLexiconEntry": [Function],
      "moveBackToWordBank": [Function],
      "setToolSettings": [Function],
      "showPopover": [Function],
    }
  }
  api={
    Object {
      "appLanguage": "en",
      "closeLoading": [Function],
      "readGlobalToolData": [Function],
      "showDialog": [Function],
      "showLoading": [Function],
      "sourceChapter": Object {},
      "targetChapter": Object {},
      "writeGlobalToolData": [Function],
    }
  }
  appLanguage="en"
  contextIdReducer={Object {}}
  currentToolViews={Object {}}
  projectDetailsReducer={Object {}}
  resourcesReducer={
    Object {
      "bibles": Array [],
      "lexicons": Object {},
    }
  }
  selectionsReducer={
    Object {
      "selections": Array [
        Object {
          "text": "text",
        },
      ],
    }
  }
  settingsReducer={
    Object {
      "toolsSettings": Object {
        "ScripturePane": Object {
          "currentPaneSettings": Array [],
        },
      },
    }
  }
  translate={[Function]}
  wordAlignmentReducer={
    Object {
      "alignmentData": Object {},
    }
  }
>
  <Tool
    actions={
      Object {
        "getWordListForVerse": [Function],
        "loadLexiconEntry": [Function],
        "moveBackToWordBank": [Function],
        "setToolSettings": [Function],
        "showPopover": [Function],
      }
    }
    api={
      Object {
        "appLanguage": "en",
        "closeLoading": [Function],
        "readGlobalToolData": [Function],
        "showDialog": [Function],
        "showLoading": [Function],
        "sourceChapter": Object {},
        "targetChapter": Object {},
        "writeGlobalToolData": [Function],
      }
    }
    appLanguage="en"
    contextIdReducer={Object {}}
    currentToolViews={Object {}}
    projectDetailsReducer={Object {}}
    resourcesReducer={
      Object {
        "bibles": Array [],
        "lexicons": Object {},
      }
    }
    selectionsReducer={
      Object {
        "selections": Array [
          Object {
            "text": "text",
          },
        ],
      }
    }
    settingsReducer={
      Object {
        "toolsSettings": Object {
          "ScripturePane": Object {
            "currentPaneSettings": Array [],
          },
        },
      }
    }
    translate={[Function]}
    wordAlignmentReducer={
      Object {
        "alignmentData": Object {},
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <TestContextContainer
        actions={
          Object {
            "getWordListForVerse": [Function],
            "loadLexiconEntry": [Function],
            "moveBackToWordBank": [Function],
            "setToolSettings": [Function],
            "showPopover": [Function],
          }
        }
        api={
          Object {
            "appLanguage": "en",
            "closeLoading": [Function],
            "readGlobalToolData": [Function],
            "showDialog": [Function],
            "showLoading": [Function],
            "sourceChapter": Object {},
            "targetChapter": Object {},
            "writeGlobalToolData": [Function],
          }
        }
        appLanguage="en"
        contextIdReducer={Object {}}
        currentToolViews={Object {}}
        projectDetailsReducer={Object {}}
        resourcesReducer={
          Object {
            "bibles": Array [],
            "lexicons": Object {},
          }
        }
        selectionsReducer={
          Object {
            "selections": Array [
              Object {
                "text": "text",
              },
            ],
          }
        }
        settingsReducer={
          Object {
            "toolsSettings": Object {
              "ScripturePane": Object {
                "currentPaneSettings": Array [],
              },
            },
          }
        }
        translate={[Function]}
        wordAlignmentReducer={
          Object {
            "alignmentData": Object {},
          }
        }
      >
        <DragDropContext(Connect(Container))
          actions={
            Object {
              "getWordListForVerse": [Function],
              "loadLexiconEntry": [Function],
              "moveBackToWordBank": [Function],
              "setToolSettings": [Function],
              "showPopover": [Function],
            }
          }
          api={
            Object {
              "appLanguage": "en",
              "closeLoading": [Function],
              "readGlobalToolData": [Function],
              "showDialog": [Function],
              "showLoading": [Function],
              "sourceChapter": Object {},
              "targetChapter": Object {},
              "writeGlobalToolData": [Function],
            }
          }
          appLanguage="en"
          contextIdReducer={Object {}}
          currentToolViews={Object {}}
          projectDetailsReducer={Object {}}
          resourcesReducer={
            Object {
              "bibles": Array [],
              "lexicons": Object {},
            }
          }
          selectionsReducer={
            Object {
              "selections": Array [
                Object {
                  "text": "text",
                },
              ],
            }
          }
          settingsReducer={
            Object {
              "toolsSettings": Object {
                "ScripturePane": Object {
                  "currentPaneSettings": Array [],
                },
              },
            }
          }
          translate={[Function]}
          wordAlignmentReducer={
            Object {
              "alignmentData": Object {},
            }
          }
        >
          <Connect(Container)
            actions={
              Object {
                "getWordListForVerse": [Function],
                "loadLexiconEntry": [Function],
                "moveBackToWordBank": [Function],
                "setToolSettings": [Function],
                "showPopover": [Function],
              }
            }
            api={
              Object {
                "appLanguage": "en",
                "closeLoading": [Function],
                "readGlobalToolData": [Function],
                "showDialog": [Function],
                "showLoading": [Function],
                "sourceChapter": Object {},
                "targetChapter": Object {},
                "writeGlobalToolData": [Function],
              }
            }
            appLanguage="en"
            contextIdReducer={Object {}}
            currentToolViews={Object {}}
            projectDetailsReducer={Object {}}
            resourcesReducer={
              Object {
                "bibles": Array [],
                "lexicons": Object {},
              }
            }
            selectionsReducer={
              Object {
                "selections": Array [
                  Object {
                    "text": "text",
                  },
                ],
              }
            }
            settingsReducer={
              Object {
                "toolsSettings": Object {
                  "ScripturePane": Object {
                    "currentPaneSettings": Array [],
                  },
                },
              }
            }
            translate={[Function]}
            wordAlignmentReducer={
              Object {
                "alignmentData": Object {},
              }
            }
          >
            <Container
              actions={
                Object {
                  "getWordListForVerse": [Function],
                  "loadLexiconEntry": [Function],
                  "moveBackToWordBank": [Function],
                  "setToolSettings": [Function],
                  "showPopover": [Function],
                }
              }
              alignTargetToken={[Function]}
              alignedTokens={Array []}
              api={
                Object {
                  "appLanguage": "en",
                  "closeLoading": [Function],
                  "readGlobalToolData": [Function],
                  "showDialog": [Function],
                  "showLoading": [Function],
                  "sourceChapter": Object {},
                  "targetChapter": Object {},
                  "writeGlobalToolData": [Function],
                }
              }
              appLanguage="en"
              clearState={[Function]}
              contextIdReducer={Object {}}
              currentToolViews={Object {}}
              loadChapterAlignments={[Function]}
              moveSourceToken={[Function]}
              projectDetailsReducer={Object {}}
              resetVerse={[Function]}
              resourcesReducer={
                Object {
                  "bibles": Array [],
                  "lexicons": Object {},
                }
              }
              selectionsReducer={
                Object {
                  "selections": Array [
                    Object {
                      "text": "text",
                    },
                  ],
                }
              }
              settingsReducer={
                Object {
                  "toolsSettings": Object {
                    "ScripturePane": Object {
                      "currentPaneSettings": Array [],
                    },
                  },
                }
              }
              sourceTokens={Array []}
              targetTokens={Array []}
              translate={[Function]}
              unalignTargetToken={[Function]}
              verseAlignments={Array []}
              verseIsValid={true}
              wordAlignmentReducer={
                Object {
                  "alignmentData": Object {},
                }
              }
            />
          </Connect(Container)>
        </DragDropContext(Connect(Container))>
      </TestContextContainer>
    </Provider>
  </Tool>
</DragDropContext(Tool)>
`;
